powerauth-java-server/src/main/java/com/wultra/security/powerauth/app/server/service/crypto/v4/CryptographyServiceHybrid.java

```diff
@@ -60,18 +153,80 @@ public SecretKey generateSharedSecretKey(ActivationRecordEntity activation) thro
+            serverPublicKeys.storePublicKey(SharedSecretAlgorithm.EC_P384_ML_L3, KeyType.MLDSA, serverKeyPairPqc.getPublic());
+            activation.setServerPublicKeys(publicKeysConverter.toDBValue(serverPublicKeys));
+
+            // Store server private key in JSON format
+            final PrivateKeyRegistry serverPrivateKeys = new PrivateKeyRegistry();
+            serverPrivateKeys.storePrivateKey(SharedSecretAlgorithm.EC_P384_ML_L3, KeyType.ECDSA, serverKeyPairEc.getPrivate());
+            serverPrivateKeys.storePrivateKey(SharedSecretAlgorithm.EC_P384_ML_L3, KeyType.MLDSA, serverKeyPairPqc.getPrivate());
+            final PrivateKeys privateKeys = serverPrivateKeysConverter.toDBValue(serverPrivateKeys, activation.getUserId(), activation.getActivationId());
+            activation.setServerPrivateKeysEncryption(privateKeys.encryptionMode());
+            activation.setServerPrivateKeys(privateKeys.privateKeysBase64());
+        } catch (CryptoProviderException | GenericCryptoException e) {
+            logger.error("Could not generate keypair", e);
+            throw localizationProvider.buildExceptionForCode(ServiceError.GENERIC_CRYPTOGRAPHY_ERROR);
+        }
     }
 
     @Override
-    public BasePublicKey convertDevicePublicKey(byte[] devicePublicKey) throws GenericServiceException {
-        // TODO
+    public BasePublicKey convertDevicePublicKey(KeyType keyType, byte[] devicePublicKey) throws GenericServiceException {
+        try {
+            switch (keyType) {
+                case ECDSA -> {
+                    final PublicKey convertedPublicKey = KEY_CONVERTOR_EC.convertBytesToPublicKey(EcCurve.P384, devicePublicKey);
+                    return EcPublicKey.builder().ecPublicKey(convertedPublicKey).build();
+                }
+                case MLDSA -> {
+                    final PublicKey convertedPublicKey = KEY_CONVERTOR_PQC_DSA.convertBytesToPublicKey(devicePublicKey);
+                    return PqcPublicKey.builder().pqcPublicKey(convertedPublicKey).build();
+                }
+            }
+        } catch (InvalidKeySpecException e) {
+            logger.error("Invalid device public key", e);
+            throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_KEY_FORMAT);
+        } catch (CryptoProviderException | GenericCryptoException e) {
+            logger.error("Key conversion failed", e);
+            throw localizationProvider.buildExceptionForCode(ServiceError.GENERIC_CRYPTOGRAPHY_ERROR);
+        }
         return null;
     }
 
     @Override
     public void storeDevicePublicKey(ActivationRecordEntity activation, BasePublicKey devicePublicKey) throws GenericServiceException {
-        // TODO
+        // The device public key is stored in JSON format in column device_public_keys
+        if (devicePublicKey instanceof EcPublicKey) {
+            final PublicKey ecPublicKey = ((EcPublicKey) devicePublicKey).getEcPublicKey();
```

Comment:
You may use this feature instead
```suggestion
        if (devicePublicKey instanceof EcPublicKey ecDevicePublicKey) {
            final PublicKey ecPublicKey = ecDevicePublicKey.getEcPublicKey();
```

Reply:
Ah, yes, I used this pattern elsewhere, that's good.